srcbpatch/streamreplacer.cpp

```diff
@@ -0,0 +1,544 @@
+        }
+        cachedAmount_ = 0;
+        pNext_->DoReplacements(toProcess, true);
+        return;
+    }
+
+    if (src_[cachedAmount_] == toProcess) // check for match
+    {
+        if (++cachedAmount_ >= src_.size())
+        {// send target - do replacement
+            for (size_t q = 0; q < trg_.size(); ++q) { pNext_->DoReplacements(trg_[q], false); }
+            cachedAmount_ = 0;
+        }
+        return;
+    }
+
+    // here:   toProcess is not our char
+    //         lets check for fast track (255/256 probability)
+    if (0 == cachedAmount_)
+    {
+        pNext_->DoReplacements(toProcess, false);
+        return;
+    }
+
+    // here:   We have some cached data
+    //         at least 1 char need to be send further
+    //         remaining cached data including toProcess need to be reprocessed for match
+
+    memcpy(buffer_.data(), src_.data(), cachedAmount_);
+    buffer_[cachedAmount_++]= toProcess;
+    size_t i = 0;
+    do
+    {
+        pNext_->DoReplacements(buffer_[i++], false); // send 1 byte after another
+    } while (0 != memcmp(src_.data(), &buffer_[i], --cachedAmount_));
+    // Everything that was needed has already been sent
+    // cachedAmount_ is zero or greater
+}
+
+
+/// <summary>
+///  creates replacer
+/// </summary>
+/// <param name="src">what we are going to replace</param>
+/// <param name="trg">this is the result of the replacement</param>
+/// <returns>Replacer for building replacement chain</returns>
+static unique_ptr<StreamReplacer> CreateSimpleReplacer(
+    unique_ptr<AbstractBinaryLexeme>& src,  // what to replace
+    unique_ptr<AbstractBinaryLexeme>& trg)  // with what
+{
+    return unique_ptr<StreamReplacer>(new UsualReplacer(src, trg));
+}
+
+
+//--------------------------------------------------
+///
+///      |--SRC 1  TRG 1  |
+///  O - |-- ...          | - o
+///      |--SRC N  TRG N  |
+/// 
+class ChoiceReplacer final : public ReplacerWithNext
+{
+    typedef struct
+    {
+        span<const char> src_;
+        span<const char> trg_;
+    }ChoiceReplacerPair;
+
+public:
+    /// <summary>
+    ///   creating ChoiceReplacer from provided pairs
+    /// </summary>
+    /// <param name="choice">vector os source & target pairs</param>
+    ChoiceReplacer(StreamReplacerChoice& choice)
+    {
+        size_t bufferSize = 0; // to allocate buffer
+        const size_t sz = choice.size();
+        rpairs_.resize(sz);
+        for (size_t i = 0; i < sz; ++i)
+        {
+            auto& vPair = choice[i]; // copy from
+            auto& rpair = rpairs_[i];// copy to
+
+            rpair.src_ = vPair.first->access();
+            const size_t sourceSize = rpair.src_.size();
+            if (bufferSize < sourceSize)
+            {
+                bufferSize = sourceSize; // calculate necessary buffer size
+            }
+
+            rpair.trg_ = vPair.second->access();
+        }
+
+        buffer_.resize(bufferSize + 1); // +1 needed for inner logic but never being accessed
```

Comment:
Looks like some hack. Do we really need it? For me it looks like 
```
DoSomeMagic(); // needed for inner logic
```

I need investigate the further code to find the answer why do we need it.

Reply:
I have no idea. I had a look on ChoiceReplacer class, tried to understand why do we need bufferSize + 1 and haven't found the answer... May be i missed smth

Reply:
This code gives access violation in case we have no this extra byte.
shift_left(buffer_.data(), buffer_.data() + cachedAmount_, szProcessed);

Reply:
Hmmmmm... Is cachedAmount_ can be greater than bufferSize by 1? Did you added +1 for this case?